Modify GET /api/user endpoint to only work for admins/group managers

adam-james-v opened the pull request:
Restrict the endpoint to only give user list when the current user is an admin.
Addresses: https://github.com/metabase/metabase-private/issues/48

adam-james-v committed:
Modify GET /api/user endpoint to only work for admins/group managers

adam-james-v committed:
Restrict group managers to only see users of groups they manage

adam-james-v committed:
Adjust segmented user tests and dedupe when user in multiple groups

adam-james-v committed:
Revert a small change that broke other tests. Change wasn't strictly needed anyway.

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48

adam-james-v committed:
Properly enable premium feature :advanced-permissions in group-manager test

adam-james-v committed:
Fix sandbox user list tests. Count query needs distinct now bc of how group manager user list is returned

adam-james-v committed:
Add back the filter I forgot to include in the count query

adam-james-v committed:
Unused require due to a change in one of the tests

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48

noahmoss commented on test/metabase/api/user_test.clj:
```diff
@@ -60,34 +60,36 @@
+    (testing "Check that group-managers can get a list of that group's users."
+      (premium-features-test/with-premium-features #{:advanced-permissions}
+        (t2.with-temp/with-temp
+            [:model/PermissionsGroup           {group-id1 :id} {:name "Test Group"}
+             :model/PermissionsGroupMembership _ {:user_id (mt/user->id :rasta) :group_id group-id1 :is_group_manager true}
+             :model/PermissionsGroupMembership _ {:user_id (mt/user->id :lucky) :group_id group-id1 :is_group_manager false}]
```
Two more test cases, pursuant to my comment above:
* A manager of one group can't list users that are in a different group which they are a member, but not a manager
* A manager of one group can't list users in groups which they are not a member

adam-james-v replied:
added tests to cover these cases, thanks for pointing them out :)

noahmoss reviewed (changes requested):
I think there's a small correctness issue around group managers (if I'm understanding the expected behavior correctly). Once that's fixed I'm happy to stamp 👍

adam-james-v committed:
Fix group-manager permissions logic on the user endpoint

adam-james-v commented on src/metabase/api/user.clj:
```diff
@@ -177,27 +187,45 @@
-                   api/*is-group-manager?*)
-               (t2/hydrate :group_ids))
-     :total  (t2/count User (filter-clauses-without-paging clauses))
+        manager-group-ids   (set (group-ids-for-user api/*current-user-id*))
```
Great catch. I've added a clause to ensure that only group-ids for groups that the user is manager of are returned. 
I've adjusted the logic in the endpoint.

adam-james-v commented on test/metabase/api/user_test.clj:
```diff
@@ -60,34 +60,36 @@
-    (testing "Check that anyone can get a list of all active Users"
+    (testing "Check that admins can get a list of all active Users"
       (mt/with-non-admin-groups-no-root-collection-perms
-        (is (= [{:id          (mt/user->id :crowberto)
-                 :email       "crowberto@metabase.com"
-                 :first_name  "Crowberto"
-                 :last_name   "Corv"
-                 :common_name "Crowberto Corv"}
```
Since the endpoint now only works for admins, the admins can see the entirety of a user entity, so it felt unnecessary to check for equivalence of the whole map. What I've done is added an assertion that the keys returned are what we expect for admins to be able to see, then just check that the correct users are returned by checking their emails, as I had done before.

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -160,9 +160,30 @@
+  "Given a `user-id` return a list of group-ids of which the user is a group manager."
+  [user-id]
+  (map :group_id
+       (t2/query
+        {:select-distinct [:permissions_group_membership.group_id]
+         :from  [:permissions_group_membership]
+         :where [:and [:= :permissions_group_membership.user_id user-id]
+                 [:= :permissions_group_membership.is_group_manager true]
+                 [:not= :permissions_group_membership.group_id (:id (perms-group/all-users))]]})))
```
nit: think it's slightly more idiomatic/readable to define an alias for the table name (e.g. `:from  [[:permissions_group_membership :pgm]]`) and use it in the rest of the query

noahmoss replied:
also nit: map is partially aligned

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -160,9 +160,30 @@
+(defn- group-ids-for-manager
+  "Given a `user-id` return a list of group-ids of which the user is a group manager."
+  [user-id]
+  (map :group_id
+       (t2/query
```
nit: slightly nicer to use `t2/select-fn-set` here, with `:group_id` as the function, to return a set of group IDs directly

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -177,27 +198,42 @@
-   include_deactivated    (s/maybe su/BooleanString)}
-  (when (or status include_deactivated)
-    (validation/check-group-manager))
+    {status              (s/maybe s/Str)
+     query               (s/maybe s/Str)
+     group_id            (s/maybe su/IntGreaterThanZero)
+     include_deactivated (s/maybe su/BooleanString)}
```
nit: indentation

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -160,9 +160,30 @@
+                 [:= :permissions_group_membership.is_group_manager true]
+                 [:not= :permissions_group_membership.group_id (:id (perms-group/all-users))]]})))
+
+(defn- check-manager-of-group
```
Could this be moved to `metabase.api.common.validation` since we already have `check-group-manager` there? Maybe the existing function could also take an optional group-id or something, or this could just be a separate helper

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -160,9 +160,30 @@
+  (api/check-403
+   (or
+    api/*is-superuser?*
+    ((set (group-ids-for-manager api/*current-user-id*)) group-id))))
```
For checking if a user is a manager of a specific group I don't think it's necessary to read their entire group list. You could just have a query that checks if a row exists for the user ID and group ID where `is_group_manager` is `true`

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -160,9 +160,30 @@
 (api/defendpoint-schema GET "/"
-  "Fetch a list of `Users`. By default returns every active user but only active users.
+  "Fetch a list of `Users` for admins or group managers.
+  By default returns only active users for admins and only active users within groups that the group manager is managing for group-managers.
```
nit (consistency)
```suggestion
  By default returns only active users for admins and only active users within groups that the group manager is managing for group managers.
```

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -177,27 +198,42 @@
+     query               (s/maybe s/Str)
+     group_id            (s/maybe su/IntGreaterThanZero)
+     include_deactivated (s/maybe su/BooleanString)}
+  (validation/check-group-manager)
```
Think we can skip this generic check if `group_id` is provided, since `check-manager-of-group` is already checking whether the user is a group manager. (But as I mentioned above, it would be nice if all this behavior could be rolled into one function)

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -177,27 +198,42 @@
-               (t2/hydrate :group_ids))
-     :total  (t2/count User (filter-clauses-without-paging clauses))
+        manager-group-ids   (set (group-ids-for-manager api/*current-user-id*))
+        group-id-clause     (cond
+                              ;; superuser can see anything, so group_id clause is ok
+                              (and api/*is-superuser?* group_id) [group_id]
+                              ;; otherwise, if the user is a group manager, only show them users in the groups they manage
+                              api/*is-group-manager?*            (vec manager-group-ids))
```
Think this logic is still slightly off. What if the user is a group manager _and_ `group_id` is passed in? I think if `group_id` is passed in we want add the filter unconditionally at this point (since we've already very that the user is either an admin or manager of the group), and if it's not passed in, we only want to add a filter for group managers.

```suggestion
        group-id-clause     (cond
                              ;; superuser can see anything, so group_id clause is ok
                              group_id [group_id]
                              ;; otherwise, if the user is a group manager, only show them users in the groups they manage
                              api/*is-group-manager?* (vec manager-group-ids))
```

noahmoss reviewed (changes requested):
Looks better! Requesting changes again for the one remaining logic error that I think is there. Everything else is nits/refactor suggestions.

adam-james-v committed:
Update src/metabase/api/user.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

adam-james-v committed:
Changed endpoint's logic and impl slightly

 - cleaner group-ids-for-manager query
 - use existing validation/check-manager-of-group
 - simplify group-id-clause since at that point we know that the user is admin or group manager

Added a test case to ensure group-manager still only gets their groups users when group_id is not supplied

These scenarious should be accounted for in this setup now

;; ADMIN GM group_id  GM of group_id OUTCOME
;; t     t    t            t         can see users from that group
;; t     t    t            f         can see users from that group (bc admin can see)
;; t     f    t            f         can see users from that group (bc admin can see)
;; t     t    f            f         can see all (admin can see all)
;; f     t    f            f         can see users from all groups they manage
;; f     t    t            t         can see users from that group
;; f     t    t            f         cannot see (bc non admin, GM of different group)

adam-james-v committed:
Merge branch 'bugfix-user-api-leaking-users-p48' of https://github.com/metabase/metabase into bugfix-user-api-leaking-users-p48

adam-james-v committed:
Alter tests to correctly check 403 in OSS setting

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48

adam-james-v committed:
Modify test to use generic verified by text since User endpoint is now more restricted.

adam-james-v commented on src/metabase/api/user.clj:
```diff
@@ -177,27 +188,44 @@
+   api/*is-superuser?*
+   (if group_id
+     (validation/check-manager-of-group group_id)
+     (validation/check-group-manager)))
```
Both of these validation functions have a default `require-admin` as true. This flag is only relevant in the OSS scenario (so the flag should be thought of as `require-admin-in-oss`). If the user is an admin, we never have to check the group manager stuff, admins can see it. 

Otherwise, we DO have to check if the group manager is a manager of the given group_id OR if they're a manager at all when no ID is given.

Here, we rely on the `require-admin` flag to ensure that OSS doesn't slip through this. Not sure it could ever happen where a user was a group manager in an Enterprise edition, but the whole instance is downgraded to OSS, so the feature is gone but the group_manager status is still there... in any case, this will prevent that user from getting user data.

adam-james-v commented on src/metabase/api/user.clj:
```diff
@@ -177,27 +188,44 @@
+                              ;; We know that the user is either admin or group manager of the given group_id (if it exists)
+                              group_id                [group_id]
+                              ;; otherwise, if the user is a group manager, only show them users in the groups they manage
+                              api/*is-group-manager?* (vec manager-group-ids))
```
Here, since we checked admin/group-manager previously, we know that the user is validated, so we can simply pass a group_id if it exists. Otherwise, if its a group-manager, we do have to restrict the users they see to be only those users which are part of whatever group(s) the manager manages. If the user is admin and there's no group_id this clause is `nil`, allowing the admin to see the full list of users.

adam-james-v commented on e2e/test/scenarios/organization/content-verification.cy.spec.js:
```diff
@@ -176,7 +171,7 @@ describeEE("scenarios > premium > content verification", () => {
         questionInfoButton().click();
         cy.findByTestId("sidebar-right")
-          .findAllByText(`${adminFullName} verified this`)
+          .findAllByText(`A moderator verified this`)
```
Since the endpoint is locked down for non-admins, the name of the verifier can't be grabbed when the user is not an admin, so the more generic message is shown instead. I think this is sensible behaviour, so I just changed this e2e test to search for that string instead.

noahmoss replied:
FYI @luizarakaki

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48

adam-james-v committed:
Strange pulse e2e test. Intercept the user api call to circumvent for now

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -177,27 +187,45 @@
+                              ;; otherwise, if the user is a group manager, only show them users in the groups they manage
+                              api/*is-group-manager?*                                    (vec manager-group-ids))
+        clauses             (user-clauses status query group-id-clause include_deactivated)]
+    {:data (cond-> (t2/select
```
this isn't fixed, FYI — I'm referring to the alignment of values associated with `:data`, `:total`, etc (not a big deal at all but it sticks out when it's inconsistent)

adam-james-v replied:
Hmm... sorry. Looks like my indentation rules are wonky or something. Won't worry about it at the moment, but I'll fix stuff on my end and have this work nicer in the future.

noahmoss reviewed (approved):
awesome, lgtm

JesseSDevaney committed:
request recipients through /api/user/recipients instead of /api/user

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48

adam-james-v committed:
I don't even know anymore. Why did this change now?

Seems that there's a collection ID where before there was a null,
which caused redirect to collection/root

adam-james-v committed:
Merge branch 'bugfix-user-api-leaking-users-p48' of https://github.com/metabase/metabase into bugfix-user-api-leaking-users-p48

adam-james-v committed:
Merge branch 'master' into bugfix-user-api-leaking-users-p48
